request location permission when location permission is denied

lqchuong opened the pull request:
request location permission when location permission is denied

nanhtrangst committed:
request location permission when location permission is denied

suesitran commented on lib/bloc/location/location_bloc.dart:
```diff
@@ -15,21 +17,104 @@ class LocationBloc extends Bloc<LocationEvent, LocationState> {
-      final LocationDetails data = await _locationService.getCurrentLocation();
-      final String dateDisplay =
-          DateFormat('dd-MMM-yyyy').format(DateTime.now());
+    final bool isConnected = await InternetConnectionChecker().hasConnection;
```
Internet connection is not needed to get user's current location. Please remove this part

suesitran commented on lib/bloc/location/location_bloc.dart:
```diff
@@ -15,21 +17,104 @@ class LocationBloc extends Bloc<LocationEvent, LocationState> {
+    //internet connected
+    if (isConnected) {
+      SharedPreferences prefs = await SharedPreferences.getInstance();
+      final bool isLoaded = prefs.getBool('isLoadedDialog') ?? false;
```
When permission is denied, we can still show permission request dialog, therefore you do not need to create a Boolean to keep track if you have showed permission request dialog previously. Just use PermissionStatus to determine if we can show permission request dialog, or if we need to use default location.

suesitran commented on lib/bloc/location/location_bloc.dart:
```diff
@@ -15,21 +17,104 @@ class LocationBloc extends Bloc<LocationEvent, LocationState> {
-      emit(LocationPermissionNotGrantedState());
-    } on Exception catch (_) {
-      emit(UnknownLocationErrorState());
+  Future<void> _requestPermissionLocationEvent(
```
This function is good, in fact it can replace the method to request current location, because it has all logic needed to request current location.

suesitran commented on lib/features/home/home_screen.dart:
```diff
@@ -52,10 +56,47 @@ class IDHome extends Screen {
           }
 
+          if (state is LocationPermissionNotGrantedState) {
+            WidgetsBinding.instance.addPostFrameCallback((timeStamp) async {
```
Instead is using async here, you can send Function callback to your dialog to preform task based on which action user chooses

suesitran commented on lib/utils/dialogs/dialog.dart:
```diff
@@ -0,0 +1,58 @@
+import 'package:flutter/material.dart';
+
+class DialogRequest extends StatefulWidget {
```
Please move this to packages/nartus_ui_package/, create a new directory and name it 'dialog'

lqchuong replied:
done.

suesitran commented on lib/utils/dialogs/dialog.dart:
```diff
@@ -0,0 +1,58 @@
+    switch (widget.typeDialog) {
+      case 0:
+        return AlertDialog(
+          title: const Text('Location Permission not granted'),
+          content: const Text(
+              'Location Permission is needed to use this app. Please Allow Interactive Diary to access location in the next dialog'),
+          actions: <Widget>[
+            FlatButton(
+              child: const Text('Allow'),
+              onPressed: () async {
+                allow = true;
+                Navigator.of(context, rootNavigator: true).pop(allow);
+              },
+            ),
+            FlatButton(
+              child: const Text('Continue'),
+              onPressed: () {
+                Navigator.of(context, rootNavigator: true).pop();
+                debugPrint('Location Denied once');
+              },
+            ),
```
This implementation does not allow the dialog to be reused later. Instead, just accept a String as title, and another String as content, with list of Actions. That way, it can be more flexible and easy to reuse.

suesitran commented on pubspec.yaml:
```diff
@@ -27,18 +27,25 @@ environment:
   google_maps_flutter: ^2.1.10
   flutter_bloc: ^8.0.1
   intl: ^0.17.0
+  shared_preferences:
```
We do not need this package yet

suesitran commented on pubspec.yaml:
```diff
@@ -27,18 +27,25 @@ environment:
   nartus_authentication:
     path: packages/nartus_authentication
-
+  internet_connection_checker: ^0.0.1+4
```
Please do not use this package. There is another story in which we'll need connectivity check, and we'll use connectivity_plus, which is a Flutter Favourite package.

suesitran reviewed (commented):
Looks good, added a few comments. And 1 questions: where are the test?

nanhtrangst committed:
- Remove Internet connection
- remove Boolean to keep track if you have showed permission request dialog previously.
- move dialog to packages/nartus_ui_package/, create a new directory and name it 'dialog'
- modify dialog widget, add parrams to it
- remove shared_preferences package and internet_connection_checker package

nanhtrangst committed:
- add test request permission
- fix alert dialog

suesitran commented on lib/bloc/location/location_bloc.dart:
```diff
@@ -32,4 +39,37 @@ class LocationBloc extends Bloc<LocationEvent, LocationState> {
+        emit(LocationPermissionDeniedState());
+        break;
+      default:
+    }
```
it's better to update location_service's `Future<PermissionStatus> requestPermission` to return our own enum, and combine `PermissionStatus.denied` and `PermissionStatus.deniedForever` into 1 `denied` status.

that way, we do not have to import `package:location/location.dart` in our main bloc, and violating [`depend_on_referenced_packages`](https://dart-lang.github.io/linter/lints/depend_on_referenced_packages.html)

suesitran commented on lib/bloc/location/location_bloc.dart:
```diff
@@ -32,4 +39,37 @@ class LocationBloc extends Bloc<LocationEvent, LocationState> {
+  Future<void> _defaultLocationServiceEvent(Emitter<LocationState> emit) async {
+    //default location when press continue button in permission dialog
+    final LocationDetails dataDefault =
+        LocationDetails(10.7840007, 106.7034988);
```
these lat/lng value can be defined as `const`

suesitran commented on lib/bloc/location/location_event.dart:
```diff
@@ -2,6 +2,11 @@ part of 'location_bloc.dart';
 
 class RequestLocationServiceEvent extends LocationEvent {}
+
+class DefaultLocationServiceEvent extends LocationEvent {}
```
all logic can be within `location_bloc`, so there's no need to have `DefaultLocationServiceEvent`

suesitran commented on lib/bloc/location/location_state.dart:
```diff
@@ -2,12 +2,18 @@ part of 'location_bloc.dart';
 
 abstract class LocationState {}
 
+class InternetDisconectedState extends LocationState {}
```
do we need this state? there's no dependent on Internet connection

suesitran commented on lib/bloc/location/location_state.dart:
```diff
@@ -2,12 +2,18 @@ part of 'location_bloc.dart';
 
 class LocationPermissionNotGrantedState extends LocationState {}
 
+class LocationPermissionDeniedState extends LocationState {}
```
`LocationPermissionDeniedState` and `LocationPermissionNotGrantedState` are the same meaning right? can u reuse `LocationPermissionNotGrantedState`, or change it to be `LocationPermissionDeniedState`

suesitran commented on lib/features/home/home_screen.dart:
```diff
@@ -3,11 +3,19 @@ import 'package:flutter_bloc/flutter_bloc.dart';
 import 'package:interactive_diary/constants/dimens.dart';
 import 'package:nartus_ui_package/nartus_ui.dart';
-
+import 'package:nartus_location/nartus_location.dart';
```
unused import

suesitran commented on lib/features/home/home_screen.dart:
```diff
@@ -3,11 +3,19 @@ import 'package:flutter_bloc/flutter_bloc.dart';
+  primaryButtonPress(BuildContext context) {}
+
+  secondaryButtonPress(BuildContext context) {
+    Navigator.of(context, rootNavigator: true).pop(false);
```
do not pop to root, we only need to dismiss the dialog

suesitran commented on lib/features/home/home_screen.dart:
```diff
@@ -52,10 +60,53 @@ class IDHome extends Screen {
+          }
+
           if (state is LocationInitial) {
+            print('RequestCurrentLocationEvent');
```
avoid print in production, please use `debugPrint()`

nanhtrangst committed:
- fix logic show diablog in location_bloc
- add enum PermissionStatusDiary

nanhtrangst committed:
fix out of date code

suesitran committed:
Merge branch 'dev' into koc/request-location-permission
